### ISSUE TITLE
Misc. bug: quantized KV cache (q5_x, q4_x) with -fa on silently fallback CPU attn

### ISSUE THREAD

**Issue by AhBumm** [ggml-org/llama.cpp#28455]
$\textcolor{red}{\textbf{(USER WAS BANNED FOR THIS POST)}}$

### Name and Version

```
$ llama-server --version
version: 0.3.0-dev (build 10791, commit 95ef7fc16)
built with GNU 13.3.0 for Linux x86_64
```

Reproduced on this build; **also verified still present on latest master `6a1a922d2` (b10819)** — `ggml_cuda_fattn_kv_type_supported()` and the default FA kernel set in `ggml/src/ggml-cuda/fattn.cu` are unchanged.

### Operating systems

- Linux (Ubuntu 24.04)

### Which llama.cpp modules do you know to be affected?

- libllama (core library)
- llama-server
- llama-cli
- llama-bench
- ggml (ggml-cuda backend / scheduler)

### Command line

```bash
llama-server -m Qwen3.8-27B-Q5_K_XL.gguf \
  -ngl 99 -fa on -c 131072 -np 4 \
  --cache-type-k q5_1 --cache-type-v q5_1 \
  --spec-type draft-mtp --spec-draft-n-max 3 \
  --host 0.0.0.0 --port 12345
```

Minimal reproduction with llama-bench:

```bash
llama-bench -m model.gguf -fa 1 -ctk q5_1 -ctv q5_1 -p 512 -n 128 -ngl 99 -t 12
```

### Problem description & steps to reproduce

With a **default build** (`GGML_CUDA_FA_ALL_QUANTS=OFF`), using KV cache quantization types **q5_0 / q5_1** (or mixed K/V combos) together with `-fa on` silently moves the entire `GGML_OP_FLASH_ATTN_EXT` operator to the **CPU backend**. The user gets no error and no warning; performance collapses.

#### Root cause

The default FA vec-kernel instantiation set in `ggml/src/ggml-cuda/fattn.cu` only contains:

```cpp
#else
    FATTN_VEC_CASES_ALL_D(GGML_TYPE_F16,  GGML_TYPE_F16)
    FATTN_VEC_CASES_ALL_D(GGML_TYPE_Q4_0, GGML_TYPE_Q4_0)
    FATTN_VEC_CASES_ALL_D(GGML_TYPE_Q8_0, GGML_TYPE_Q8_0)
    FATTN_VEC_CASES_ALL_D(GGML_TYPE_BF16, GGML_TYPE_BF16)
#endif // GGML_CUDA_FA_ALL_QUANTS
```

`ggml_cuda_fattn_kv_type_supported()` returns `false` for `Q4_1` / `Q5_0` / `Q5_1` unless `GGML_CUDA_FA_ALL_QUANTS` is defined, so `get_best_fattn_kernel()` returns `BEST_FATTN_KERNEL_NONE`, the CUDA backend rejects the op, and the ggml scheduler assigns FlashAttention to the CPU backend — for every layer, every step. The tile kernel reads K/V as `half2` only, so it cannot take quantized KV either. `-fa 0` with quantized V fails outright (non-FA path requires a transposeable f16 V), so there is **no working GPU path at all** for these KV types without the extra build flag.

This also fully explains #27109 (q4_1 K cache prefill collapse on qwen35 hybrid): `Q4_1` sits in the same `return false` branch as `Q5_0`/`Q5_1`, while `Q4_0` and `Q8_0` are in the supported list — which matches that issue's observation that q4_0 works fine but q4_1 collapses.

#### Runtime evidence (RTX 4080, 27B model, pp512 with q5_1 KV)

```
GPU utilization:  0 %     GPU power: ~62 W
CPU usage:        ~930 %  (9+ cores saturated by llama-server/llama-bench)
```

With q8_0 KV the identical run shows GPU ≥30% and single-digit CPU %, consistent with device→host KV copies plus CPU attention.

#### K/V type isolation matrix (llama-bench pp512, -ngl 99 -fa 1 -t 12, qwen35 27B Q5_K_XL)

| K type | V type | pp512 (t/s) | vs q8_0/q8_0 |
|--------|--------|------------:|-------------:|
| q8_0   | q8_0   | 1511.92     | 1.0x         |
| **q8_0** | **q5_1** | **197.05** | **7.7x slower** |
| **q5_1** | q8_0   | 427.61      | 3.5x slower  |
| f16    | **q5_1** | **190.36** | **7.9x slower** |
| **q5_1** | f16    | 483.14      | 3.1x slower  |
| **q5_1** | **q5_1** | **196.62** | **7.7x slower** |

Note `llama-bench` still prints `fa=1 q5_1` in its results table while actually measuring CPU attention + GPU↔host copies — the benchmark output itself is misleading.

#### Impact in llama-server (3.6K-token prompt, 128 gen, same machine)

| config | prompt eval | eval (decode) |
|---|---|---|
| q5_1 KV, default build | 38.4 t/s | 6.5 t/s |
| q8_0 KV, default build | 1348 t/s | 30.1 t/s |
| **q5_1 KV, `GGML_CUDA_FA_ALL_QUANTS=ON` rebuild** | **1162 t/s** | **38.5 t/s** |

The CPU-attention cost scales with KV length, so long-context workloads are hit hardest: a 3 s TTFT turned into a ~95 s TTFT, which is very easy to misdiagnose as "long context kills my GPU".

#### Related issues

- #27109 — q4_1 KV prefill collapse on qwen35 hybrid (RTX 3090), root cause unknown to the reporter; explained by this issue (q4_1 not in the default kernel set)
- #24485 — FR to make `GGML_CUDA_FA_ALL_QUANTS` default or warn at runtime (closed stale; the silent fallback keeps trapping users, see data above)
- #20866 — mixed K/V quantization cannot be offloaded to GPU (same `supported()` matrix)
- #27905 — RPC + certain KV quant types trigger FA fatal error (same matrix, different symptom)

### First Bad Commit

Not a regression of a specific commit — the default-compiled kernel set has excluded q5/q4_1 combos ever since the `GGML_CUDA_FA_ALL_QUANTS` split was introduced (#24403 later consolidated the checks into `supported()`). The core problem is that the **fallback is completely silent**.

### Relevant log output

There is no warning at all in the console — that is the point of this report. The only observable symptoms are performance collapse and GPU utilization at 0%.

<details>
<summary>Runtime evidence: pp512 with q5_1 KV (sampled via top/nvidia-smi during llama-bench)</summary>

```
GPU utilization:  0 %     GPU power: ~62 W
CPU usage:        ~930 %  (9+ cores saturated by llama-server/llama-bench)
```

With q8_0 KV the identical run shows GPU >=30% and single-digit CPU %.

</details>

<details>
<summary>llama-bench: default build vs GGML_CUDA_FA_ALL_QUANTS=ON rebuild</summary>

Default build (ALL_QUANTS=OFF):

```
| qwen35 27B Q5_K - Medium | ... | q5_1 | q5_1 | 1 | pp512 |  197.15 t/s |
| qwen35 27B Q5_K - Medium | ... | q5_1 | q5_1 | 1 | tg128 |   25.98 t/s |
```

Rebuild with -DGGML_CUDA_FA_ALL_QUANTS=ON:

```
| qwen35 27B Q5_K - Medium | ... | q5_1 | q5_1 | 1 | pp512 | 1509.70 t/s |
| qwen35 27B Q5_K - Medium | ... | q5_1 | q5_1 | 1 | tg128 |   30.55 t/s |
```

Note: llama-bench prints `fa=1 q5_1` in both tables while the first one actually measures CPU attention + GPU<->host copies — the benchmark output itself is misleading.

</details>

A `--verbose`-level line such as "flash attention for KV type q5_1 not compiled in this build; op scheduled on CPU" would have saved hours of debugging.

### Suggested fixes (any one would help)

1. Include q5_0/q5_1 (+ q4_1, + mixed K/V) in the default FA kernel set — these types are advertised as valid `-ctk/-ctv` values in the CLI help, docs, and `llama-bench`'s own type parser, so a default build should either support them or refuse them loudly.
2. Emit a startup/runtime warning when `-fa on` + quantized KV is requested but the CUDA FA kernel for that combination is not compiled in (i.e. when a FA op is scheduled onto the CPU backend).
3. At minimum: print a warning in `llama-bench` when any FA op was scheduled off-GPU, so benchmark results aren't silently comparing different code paths.

### Workaround

Rebuild with:

```bash
cmake -B build -DGGML_CUDA=ON -DCMAKE_CUDA_ARCHITECTURES=89 \
      -DGGML_CUDA_FA_ALL_QUANTS=ON
cmake --build build --config Release -j 12
```

Same q5_1/q5_1 benchmark after rebuild:

```
| qwen35 27B Q5_K - Medium | ... | q5_1 | q5_1 | 1 | pp512 | 1509.70 t/s |
| qwen35 27B Q5_K - Medium | ... | q5_1 | q5_1 | 1 | tg128 |   30.55 t/s |
```

(Alternative: use `-ctk q8_0 -ctv q8_0` or f16 KV, which are in the default kernel set, at the cost of ~60% more KV VRAM vs q5_1.)

Happy to provide further traces (perf samples, scheduler dumps) or test a fix.

**Comment by sinanezhadian** [comment]
Very clear report. One consequence worth adding, because it widens who this affects.

The silent fallback changes the **numerics**, not only the speed. CPU flash-attention and the CUDA FA kernels accumulate in different orders, so a default build running `-ctk q5_1 -ctv q5_1` is producing different logits from the same build at q8_0 — not just producing them more slowly.

That extends the harm from performance tables to quality tables. Anyone who has ranked KV cache quantization types by perplexity on a default CUDA build has been attributing to *quantization* a difference that is partly a CPU-versus-CUDA backend swap, because q4_0 and q8_0 stayed on the GPU while q5_0, q5_1 and q4_1 silently did not. The comparison people think they are running is "how much does this KV quant type cost me in quality"; the comparison they actually ran confounds that with a backend change.

**The useful part: you have already built the controlled experiment.** Your two binaries at identical `q5_1/q5_1` — default build at 197.15 t/s pp512 and the `GGML_CUDA_FA_ALL_QUANTS=ON` rebuild at 1509.70 — hold the model, the GPU and the KV quantization type fixed. The only thing that differs is which backend runs attention. Running `llama-perplexity` on both binaries at that setting isolates a pure backend effect with no confounds, which is a rare thing to have available for free.

If that delta lands at or above the perplexity differences people routinely use to choose between KV quant types, it makes your suggestion 2 a good deal stronger than the timing evidence alone does: a loud warning would then be protecting correctness comparisons, not just saving people from a slow run. It also sharpens suggestion 3 — `llama-bench` printing `fa=1 q5_1` mislabels the *numerical* configuration, not only the performance one, so the warning arguably belongs in the results table rather than behind `--verbose`.

For scale on how large backend-attributable differences can get: holding weights, decoding parameters and hardware constant and changing only the inference engine moved benchmark scores by up to 16.6 points across five engines in [arXiv:2605.19537](https://arxiv.org/abs/2605.19537). That is a cross-engine result at a different scale, so treat it as a reason to measure rather than as a prediction for this case — the point is only that "the backend changed underneath a benchmark that kept its label" is a known way to get a large number, and it is exactly what is happening here.

**Comment by JohannesGaessler** [comment]
This is a known issue.

Also according to the llama.cpp [AI usage policy](https://github.com/ggml-org/llama.cpp/blob/master/CONTRIBUTING.md#ai-usage-policy):

>It is strictly prohibited to use AI to write your posts for you (bug reports, feature requests, pull request descriptions, Github discussions, responding to humans, ...).
